src/foam/nanos/test/SmartTest.java

```diff
@@ -0,0 +1,145 @@
+package foam.nanos.test;
+
+import foam.core.X;
+import foam.core.EmptyX;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.lang.System.err;
+
+public class SmartTest extends Test {
+  public static final String TEST_METHOD_SUFFIX = "Test";
+  public static final String CLEANUP_METHOD_SUFFIX = "Cleanup";
+  public static final String SETUP_METHOD_SUFFIX = "Setup";
+
+  private Method globalSetupMethod;
+  private Method globalCleanupMethod;
+  private Method preTestMethod;
+  private Method postTestMethod;
+  private final Map<String, TestChain> testChains = new HashMap<>();
+
+  protected X x_ = EmptyX.instance();
+
+  public X getX() {
```

Comment:
What is getX() and setX() used for? I believe that Test already implements ContextAware and has these methods, but why is it needed in the first place? Having two contexts: one passed as an argument and one from getX() opens the possibility of using the wrong one.

Reply:
I used getX() and setX() to expose the X SmartTest gets from the runTest method. The X that I get from the parent's getter and setter (ContextAware) was giving problems when using sudo. This X returns every service as a null. Eg: userDAO. 

Also all the other ways to test eg GUI/beanshell test, Modelled test use, the x that is passed to the runTest which is different from the X returned from the parents' getX() method.